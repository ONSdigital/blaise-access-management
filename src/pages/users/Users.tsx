import React, { ReactElement, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { User } from "../../../Interfaces";
import { ExternalLink, ONSErrorPanel, ONSLoadingPanel } from "blaise-design-system-react-components";
import { getAllUsers } from "../../utilities/http";
import Breadcrumbs from "../../Components/Breadcrumbs";
import UsersTable from "./UsersTable";

interface Props {
    currentUser: User | undefined;
    externalCATIUrl: string;
}


function Users({ currentUser, externalCATIUrl }: Props): ReactElement {
    const [users, setUsers] = useState<User[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");
    const [listLoading, setListLoading] = useState<boolean>(true);

    useEffect(() => {
        getUserList().then(() => console.log("Call getUserList Complete"));
    }, []);

    async function getUserList() {
        setUsers([]);
        setListLoading(true);

        const [success, instrumentList] = await getAllUsers();
        setListLoading(false);

        if (!success) {
            setListError("Unable to load users.");
            return;
        }

        if (instrumentList.length === 0) {
            setListError("No installed users found.");
        }

        setUsers(instrumentList);
    }


    return <>
        <Breadcrumbs BreadcrumbList={
            [
                { link: "/", title: "Home" },
            ]
        } />

        <main id="main-content" className="page__main u-mt-no">
            <h1 className="u-mb-l">Manage users</h1>
            <ul className="list list--bare list--inline ">
                <li className="list__item ">
                    <Link to={"/users/new"}>
                        Create new user
                    </Link>
                </li>
                <li className="list__item ">
                    <Link to={"/users/upload"}>
                        Bulk upload users
                    </Link>
                </li>
            </ul>
            <p className="u-mt-m">
                <ExternalLink text={"Link to CATI dashboard"}
                    link={externalCATIUrl}
                    id={"cati-dashboard"} />
            </p>
            {listError.includes("Unable") && <ONSErrorPanel />}

            {
                listLoading ?
                    <ONSLoadingPanel />
                    :
                    <UsersTable users={users} currentUser={currentUser} listError={listError} />
            }
        </main>
    </>;
}

export default Users;
