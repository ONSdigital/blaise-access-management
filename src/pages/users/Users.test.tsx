import React from "react";
import { render, waitFor, cleanup, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { act } from "react-dom/test-utils";
import { createMemoryHistory } from "history";
import { Router } from "react-router";
import { User } from "blaise-api-node-client";
import Users from "./Users";
import { mock_server_request_Return_JSON } from "../../tests/utils";

const signedInUser: User = {
    defaultServerPark: "gusty",
    name: "TestUser123",
    role: "DST",
    serverParks: ["gusty"]
};

const userList: User[] = [
    { defaultServerPark: "gusty", name: "TestUser123", role: "DST", serverParks: ["gusty"] },
    { defaultServerPark: "gusty", name: "SecondUser", role: "BDSS", serverParks: ["gusty"] }
];

describe("Manage Users page", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, userList);
    });

    it("view users page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <Users currentUser={signedInUser} externalCATIUrl={"url/"} />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });

    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Users currentUser={signedInUser} externalCATIUrl={"url/"} />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });

        await waitFor(() => {
            expect(screen.getByText(/Manage users/i)).toBeDefined();
            expect(screen.getByText(/Create new user/i)).toBeDefined();
            expect(screen.getByText(/TestUser123/i)).toBeDefined();
            expect(screen.getByText(/SecondUser/i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});


describe("Given the API returns malformed json", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, { text: "Hello" });
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Users currentUser={signedInUser} externalCATIUrl={"url/"} />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });


        await waitFor(() => {
            expect(screen.getByText(/Sorry, there is a problem with this service. We are working to fix the problem. Please try again later./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns an empty list", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, []);
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Users currentUser={signedInUser} externalCATIUrl={"url/"} />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });


        await waitFor(() => {
            expect(screen.getByText(/No installed users found./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
