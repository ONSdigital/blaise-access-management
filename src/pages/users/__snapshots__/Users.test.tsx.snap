// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Users page view users page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="ons-breadcrumb"
      >
        <ol
          class="ons-breadcrumb__items ons-u-fs-s"
        >
          <li
            class="ons-breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="ons-breadcrumb__link"
              href="/"
            >
              Home
            </a>
            <svg
              class="ons-svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="ons-page__main ons-u-mt-no"
        id="main-content"
      >
        <h1
          class="ons-u-mb-l"
        >
          Manage users
        </h1>
        <ul
          class="ons-list ons-list--bare ons-list--inline "
        >
          <li
            class="ons-list__item "
          >
            <a
              href="/users/new"
            >
              Create new user
            </a>
          </li>
          <li
            class="ons-list__item "
          >
            <a
              href="/users/upload"
            >
              Bulk upload users
            </a>
          </li>
        </ul>
        <table
          class="ons-table ons-u-mt-m"
          id="users-table"
        >
          <thead
            class="ons-table__head "
          >
            <tr
              class="ons-table__row"
            >
              <th
                class="ons-table__header "
                scope="col"
              >
                <span>
                  Name
                </span>
              </th>
              <th
                class="ons-table__header "
                scope="col"
              >
                <span>
                  Role
                </span>
              </th>
              <th
                class="ons-table__header "
                scope="col"
              >
                <span>
                  Actions
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            class="ons-table__body"
          >
            <tr
              class="ons-table__row"
              data-testid="users-table-row"
            >
              <td
                class="ons-table__cell "
              >
                TestUser123 
                (Current user)
              </td>
              <td
                class="ons-table__cell "
              >
                DST
              </td>
              <td
                class="ons-table__cell "
              >
                <a
                  href="/users/TestUser123"
                >
                  View user
                </a>
              </td>
            </tr>
            <tr
              class="ons-table__row"
              data-testid="users-table-row"
            >
              <td
                class="ons-table__cell "
              >
                SecondUser
              </td>
              <td
                class="ons-table__cell "
              >
                BDSS
              </td>
              <td
                class="ons-table__cell "
              >
                <a
                  href="/users/SecondUser"
                >
                  View user
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumb"
      class="ons-breadcrumb"
    >
      <ol
        class="ons-breadcrumb__items ons-u-fs-s"
      >
        <li
          class="ons-breadcrumb__item"
          id="breadcrumb-0"
        >
          <a
            class="ons-breadcrumb__link"
            href="/"
          >
            Home
          </a>
          <svg
            class="ons-svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
      </ol>
    </nav>
    <main
      class="ons-page__main ons-u-mt-no"
      id="main-content"
    >
      <h1
        class="ons-u-mb-l"
      >
        Manage users
      </h1>
      <ul
        class="ons-list ons-list--bare ons-list--inline "
      >
        <li
          class="ons-list__item "
        >
          <a
            href="/users/new"
          >
            Create new user
          </a>
        </li>
        <li
          class="ons-list__item "
        >
          <a
            href="/users/upload"
          >
            Bulk upload users
          </a>
        </li>
      </ul>
      <table
        class="ons-table ons-u-mt-m"
        id="users-table"
      >
        <thead
          class="ons-table__head "
        >
          <tr
            class="ons-table__row"
          >
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                Name
              </span>
            </th>
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                Role
              </span>
            </th>
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                Actions
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="ons-table__body"
        >
          <tr
            class="ons-table__row"
            data-testid="users-table-row"
          >
            <td
              class="ons-table__cell "
            >
              TestUser123 
              (Current user)
            </td>
            <td
              class="ons-table__cell "
            >
              DST
            </td>
            <td
              class="ons-table__cell "
            >
              <a
                href="/users/TestUser123"
              >
                View user
              </a>
            </td>
          </tr>
          <tr
            class="ons-table__row"
            data-testid="users-table-row"
          >
            <td
              class="ons-table__cell "
            >
              SecondUser
            </td>
            <td
              class="ons-table__cell "
            >
              BDSS
            </td>
            <td
              class="ons-table__cell "
            >
              <a
                href="/users/SecondUser"
              >
                View user
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
